web/src/main/java/com/microsoft/applicationinsights/web/internal/httputils/HttpServerHandler.java

```diff
@@ -0,0 +1,124 @@
+package com.microsoft.applicationinsights.web.internal.httputils;
+
+import com.microsoft.applicationinsights.TelemetryClient;
+import com.microsoft.applicationinsights.common.CommonUtils;
+import com.microsoft.applicationinsights.internal.logger.InternalLogger;
+import com.microsoft.applicationinsights.telemetry.Duration;
+import com.microsoft.applicationinsights.telemetry.RequestTelemetry;
+import com.microsoft.applicationinsights.web.internal.RequestTelemetryContext;
+import com.microsoft.applicationinsights.web.internal.ThreadContext;
+import com.microsoft.applicationinsights.web.internal.WebModulesContainer;
+import org.apache.commons.lang3.Validate;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.http.annotation.Experimental;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.net.MalformedURLException;
+import java.util.Date;
+
+/**
+ * This Helper Handler class provides the required methods to instrument requests.
+ * @param <P> The HttpRequest entity
+ * @param <Q> The HttpResponse entity
+ */
+@Experimental
+public final class HttpServerHandler<P /* >>> extends @NonNull Object */, Q> {
```

Comment:
@reyang  I want to get your opinion on - should we make this class implement an interface `HttpHandler<P, Q>` with `handleStart()` and `handleEnd()` methods? Do we envision a scenario where we wrap httpClients as well and in that case this might be helpful. If not, then I think it's not needed.

Reply:
If we have generic handler that covers both server and client, probably we should avoid the name handleStart and handleEnd, as the concept `start` has very different meaning for server and client.

Reply:
I think this is not very feasible atm. Clients use different request interfaces for example https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/HttpRequest.html and sometimes these interface do not have the methods like scheme etc.

Reply:
Since we keep this API experimental for now, we can introduce interface when we feel confident.